### ISSUE TITLE
[HOLD for payment 2024-02-14] [HOLD for payment 2024-01-18] Consolidate GetDislplayName methods

### ISSUE THREAD

**Issue by puneetlath** [Expensify/App#31312]
We have a bunch of different methods that are used to get the display name: 
- PersonalDetails.getDisplayNameForTypingIndicator
- PersonalDetails.getDisplayName
- ReportUtils.getDisplayNamesStringFromTooltips
- ReportUtils.getDisplayNameForParticipant
- ReportUtils.getDisplayNamesWithTooltips
- PersonalDetailsUtils.getDisplayNameOrDefault

I would like to do two things: 
1. Conslidate these as much as possible. They probably can't all be consolidated, but I would think at least `PersonalDetailsUtils.getDisplayNameOrDefault` and `PersonalDetails.getDisplayName` and `ReportUtils.getDisplayNameForParticipant` could be
2. Make sure that we are falling back on the login if we have it. Right now `ReportUtils.getDisplayNameForParticipant` doesn't fall back on the login, even if it is available, and I would like to update it to do so

**Comment by koko57** [comment]
ready to work on this issue

**Comment by koko57** [comment]
@puneetlath I've analyzed the methods you mentioned and looking for another parts of code that have to do something with creating a displayName. For now leaving my findings for each method:
- PersonalDetails.getDisplayNameForTypingIndicator
only used in: https://github.com/Expensify/App/blob/47c7819244ccd1ec46e409c7f254791999f2637d/src/pages/home/report/ReportTypingIndicator.js#L45-L46 - we can try to use getDisplayName instead
- PersonalDetails.getDisplayName
it could be possibly merged with getDisplayNameOrDefault and getDisplayNameForParticipant
- ReportUtils.getDisplayNamesStringFromTooltips
it's only used here: https://github.com/Expensify/App/blob/47c7819244ccd1ec46e409c7f254791999f2637d/src/libs/GroupChatUtils.ts#L16-L22
and it uses `getDisplayNamesWithTooltips`. So we're creating a list of objects just to extract a display name from them 🤔. I think this logic should be simplified not to iterate over a list twice
- ReportUtils.getDisplayNameForParticipant
maybe we could replace this method, getDisplayName and getDisplayNameOrDefault and create one that will be used instead of these 3
- ReportUtils.getDisplayNamesWithTooltips
iterates over personalDetailsList array, uses `getDisplayNameForParticipant`
- PersonalDetailsUtils.getDisplayNameOrDefault
same as in getDisplayNameForParticipant

**Comment by koko57** [comment]
also `extractFirstAndLastNameFromAvailableDetails` in PersonalDetails has a very similar functionality

**Comment by koko57** [comment]
for your [PR](https://github.com/Expensify/App/pull/31275) - I think we can merge it and then I will work on mine. I would rather split the whole refactor into several smaller PRs. It would be easier to test for regressions.

**Comment by puneetlath** [comment]
@koko57 my PR was merged.

**Comment by situchan** [comment]
If dedicated C+ is needed to review all sub-PRs for this issue, I am happy to help.
@puneetlath If you agree, you can assign me on this issue.

**Comment by koko57** [comment]
@puneetlath I've opened the first draft [PR](https://github.com/Expensify/App/pull/31495). I removed the getDisplayNameForTypingIndicator method. I actually wonder if we need to check if the we have user's login or ID in reportUserIsTyping object. I tried to create a new user in OldDot and then was typing a message in NewDot it looked like we have the ID. But maybe it's for the old users? So leaving this logic now, let me know if it's still necessary.

Another thing: while working with ReportTypingIndicator (the only component that used that removed method) I noticed that we don't really need this other component https://github.com/Expensify/App/blob/8c9925d29c8f51349c742b0203a2d35ef6a651c2/src/pages/home/report/ReportTypingIndicator.js#L46-L52 - we could use the same approach as in this other case: https://github.com/Expensify/App/blob/8c9925d29c8f51349c742b0203a2d35ef6a651c2/src/pages/home/report/ReportTypingIndicator.js#L57-L63 
and it actually looks the same (screenshots - I changed for the red color the implementation with Text instead of TextWithEllipsis)

<img width="727" alt="Screenshot 2023-11-17 at 18 20 39" src="https://github.com/Expensify/App/assets/28020445/894a0bfc-6457-40af-b550-daa81703d80f">
<img width="710" alt="Screenshot 2023-11-17 at 18 23 07" src="https://github.com/Expensify/App/assets/28020445/ca4a54b8-c3a9-455e-9b5a-fa22cada136e">

I'm trying to find the reason why now we use two different components for displaying this text. Maybe I'm missing something? If it would be ok to use the Text only for both cases, we could remove TextWithEllipsis (it's only used in ReportTypingIndicator). What do you think? 🙂

**Comment by koko57** [comment]
@puneetlath and regarding the issue and the next steps:
- what do you think about moving `getDisplayName` and `extractFirstAndLastNameFromAvailableDetails` and `getCountryISO` to PersonalDetailsUtils? They are now in actions/PersonalDetails. Should we leave in the actions only the functions that write to API?
- same with  `getDisplayNameForParticipant` and other methods that deal only with allPersonalDetails object - should we move them to PersonalDetailsUtils?
- for now, I think that we can leave only `getDisplayNameForParticipant` and `getDisplayNamesWithTooltips` methods. The latter one iterates over an array uses `getDisplayNameForParticipant` and it's used in several places. I would remove `getDisplayNamesStringFromTooltips`

**Comment by puneetlath** [comment]
>If dedicated C+ is needed to review all sub-PRs for this issue, I am happy to help.

Sounds good. 

>what do you think about moving getDisplayName and extractFirstAndLastNameFromAvailableDetails and getCountryISO to PersonalDetailsUtils? They are now in actions/PersonalDetails. Should we leave in the actions only the functions that write to API?

Yes, I agree with this. Do you @situchan? 

>same with getDisplayNameForParticipant and other methods that deal only with allPersonalDetails object - should we move them to PersonalDetailsUtils?

This makes sense to me also. 


>for now, I think that we can leave only getDisplayNameForParticipant and getDisplayNamesWithTooltips methods. The latter one iterates over an array uses getDisplayNameForParticipant and it's used in several places. I would remove getDisplayNamesStringFromTooltips

Great!

**Comment by puneetlath** [comment]
> I actually wonder if we need to check if the we have user's login or ID in reportUserIsTyping object. I tried to create a new user in OldDot and then was typing a message in NewDot it looked like we have the ID. But maybe it's for the old users? So leaving this logic now, let me know if it's still necessary.

Don't we fall back on "someone is typing..." if we don't have the info? 

>Another thing: while working with ReportTypingIndicator (the only component that used that removed method) I noticed that we don't really need this other component

Nice find, I think consolidation makes sense if it isn't needed. But maybe @situchan can take a deeper look and weigh in.

**Comment by koko57** [comment]
> > I actually wonder if we need to check if the we have user's login or ID in reportUserIsTyping object. I tried to create a new user in OldDot and then was typing a message in NewDot it looked like we have the ID. But maybe it's for the old users? So leaving this logic now, let me know if it's still necessary.
> 
> Don't we fall back on "someone is typing..." if we don't have the info?

Yes, we do, but I meant checking if the key is a user ID or email login. For newly created account in the OD we do have an ID for the user, but maybe for the old accounts we only have an email. But I think we can revisit this logic later on.

For the removing TextWithEllipsis - as you asked @situchan - I'm waiting for the feedback before I will refactor this one.

For moving the methods - I would do this in a separate PR - a PR only for moving these methods.

**Comment by situchan** [comment]
> I'm trying to find the reason why now we use two different components for displaying this text. Maybe I'm missing something? If it would be ok to use the Text only for both cases, we could remove TextWithEllipsis (it's only used in ReportTypingIndicator). What do you think? 🙂

It's intentional. If display name is very long, it should be truncated at name, not on typing...

Original: `Very long display name is typing...`
Bad: `Very long display name is ...` (using `Text`)
Good: `Very long display ... is typing` (using `TextWithEllipsis`)

But this isn't necessary when multiple users are typing because "Multiple users are typing..." text is short enough to display even in small screen width.

**Comment by koko57** [comment]
@situchan great, thanks a lot for the explanation! So leaving it as it is.

**Comment by situchan** [comment]
> * what do you think about moving `getDisplayName` and `extractFirstAndLastNameFromAvailableDetails` and `getCountryISO` to PersonalDetailsUtils? They are now in actions/PersonalDetails. Should we leave in the actions only the functions that write to API?
> * same with  `getDisplayNameForParticipant` and other methods that deal only with allPersonalDetails object - should we move them to PersonalDetailsUtils?

Agree as long as it doesn't cause cyclic dependencies issue

> * for now, I think that we can leave only `getDisplayNameForParticipant` and `getDisplayNamesWithTooltips` methods. The latter one iterates over an array uses `getDisplayNameForParticipant` and it's used in several places. I would remove `getDisplayNamesStringFromTooltips`

👍

**Comment by puneetlath** [comment]
Ok nice! Are there any questions outstanding? I got a bit lost.

**Comment by koko57** [comment]
@puneetlath No, I think everything is clear now 🙂

**Comment by puneetlath** [comment]
Great!

**Comment by puneetlath** [comment]
@situchan how's the review going?

**Comment by koko57** [comment]
@puneetlath on the next PR - I will try to remove the other methods and leave the ones I mentioned:
> * for now, I think that we can leave only `getDisplayNameForParticipant` and `getDisplayNamesWithTooltips` methods. The latter one iterates over an array uses `getDisplayNameForParticipant` and it's used in several places. I would remove `getDisplayNamesStringFromTooltips`

Moving all methods concerning personalDetails to personalDetailsUtils is quite a big refactor, so I'd leave it for the last PR.

**Comment by puneetlath** [comment]
Sounds good!

**Comment by koko57** [comment]
Nice! 🙂 I'm working on my second PR

**Comment by koko57** [comment]
@puneetlath I've opened a new draft PR for this issue: https://github.com/Expensify/App/pull/32568

I've just noticed that we have a new displayName method 😅 
https://github.com/Expensify/App/blob/2feaac80198481546ee6aa630aa1c7b386ec5605/src/libs/PersonalDetailsUtils.js#L200-L210

**Comment by puneetlath** [comment]
Omg, they are spawning faster than we can squash them!

**Comment by koko57** [comment]
I'm waiting with opening my 2nd [PR](https://github.com/Expensify/App/pull/32568) for my [1st one](https://github.com/Expensify/App/pull/31495) being reviewed and merged. @situchan friendly bump 🙏🏻

**Comment by situchan** [comment]
in a few hrs

**Comment by puneetlath** [comment]
@situchan will you be able to review again today?

**Comment by situchan** [comment]
yes

**Comment by koko57** [comment]
@puneetlath @situchan I've just opened the 2nd PR for this ticket https://github.com/Expensify/App/pull/32568

- removed getDisplayNamesStringFromTooltips - we used it only in one other function to extract the displayNames from the array created by another method
- removed getDisplayNameOrDefault - we can use ReportUtils.getDisplayNameForParticipant and add fallback for "Hidden" only when we're not sure if the ID we're passing as a param is undefined; getDisplayNameOrDefault returned "Hidden" when there was no displayName and no defaultName, ReportUtils.getDisplayNameForParticipant returns empty string when ID is undefined
- renamed getDisplayName to createDisplayName - we're not getting the displayName here from the personalDetailsList, we're creating a new one for the optimistic data; also to avoid confusing it with the other methods

**Comment by koko57** [comment]
@puneetlath I see that `getDisplayNameOrDefault` has become being used more and more in the app lately and it's functionality has changed a bit since I started working on this issue. My PR removes this method, but as it's now heavily used I constantly need to fight conflicts. I wonder what do you think of this method - should we leave it as it is? Maybe be could remove `ReportUtils.getDisplayNameForParticipant` and use `getDisplayNameOrDefault` instead?

**Comment by situchan** [comment]
#32792 is good example

**Comment by koko57** [comment]
@puneetlath @situchan I closed the old PR removing `getDisplayNameOrDefault` method. I've just opened a new one https://github.com/Expensify/App/pull/33930 - with the other two changes:

- removed getDisplayNamesStringFromTooltips - we used it only in one other function to extract the displayNames from the array created by another method
- renamed getDisplayName to createDisplayName - we're not getting the displayName here from the personalDetailsList, we're creating a new one for the optimistic data; also to avoid confusing it with the other methods

I've seen that this issue was mentioned [here](https://github.com/Expensify/App/issues/33903#issuecomment-1875944100) - I can take care of this issue as a part of the refactor, but I'll be able to handle it next week (tomorrow I'm ooo). Let me know wdyt about it 🙂

**Comment by puneetlath** [comment]
That sounds good to me. As long as we are consolidating and don't have a bunch of duplicated logic that sounds good to me.

**Comment by koko57** [comment]
@puneetlath @situchan thanks for the review! I'll review once again all the changes that were done and as some other changes were introduced while I was working on these 2 PRs I will list the things that should be done in another PR (if any). If there's not other things to do here we can close the ticket. 
For this issue https://github.com/Expensify/App/issues/33903#issuecomment-1875944100 I'm waiting for retesing and confirmation if the issue still exists, bc I couldn't recreate it myself.

**Comment by puneetlath** [comment]
Sounds good!

**Comment by koko57** [comment]
SUMMARY:
> * PersonalDetails.getDisplayNameForTypingIndicator
>   only used in: https://github.com/Expensify/App/blob/47c7819244ccd1ec46e409c7f254791999f2637d/src/pages/home/report/ReportTypingIndicator.js#L45-L46
-> removed in https://github.com/Expensify/App/pull/31495 - we're using `getDisplayNameForParticipant` now

> * PersonalDetails.getDisplayName
>   it could be possibly merged with getDisplayNameOrDefault and getDisplayNameForParticipant

-> renamed to `createDisplayName` in https://github.com/Expensify/App/pull/33930 as it serves different purpose 

> * ReportUtils.getDisplayNamesStringFromTooltips
>   it's only used here: https://github.com/Expensify/App/blob/47c7819244ccd1ec46e409c7f254791999f2637d/src/libs/GroupChatUtils.ts#L16-L22
>   and it uses `getDisplayNamesWithTooltips`. So we're creating a list of objects just to extract a display name from them 🤔. I think this logic should be simplified not to iterate over a list twice

-> removed in https://github.com/Expensify/App/pull/33930  -  we're using `getDisplayNameForParticipant` now

> * ReportUtils.getDisplayNameForParticipant
>   maybe we could replace this method, `getDisplayName` and `getDisplayNameOrDefault` and create one that will be used instead of these 3

-> TODO: I need to investigate once again if we an only use one method - `getDisplayNameOrDefault` changed a lot since I started working on this issue 

> * ReportUtils.getDisplayNamesWithTooltips
>   iterates over personalDetailsList array, uses `getDisplayNameForParticipant`

-> not removing it, it stays as it is for now - after the investigation mentioned above it may require changing the method it's using

> * PersonalDetailsUtils.getDisplayNameOrDefault
>   same as in getDisplayNameForParticipant

-> same as above

other ideas:

> * moving `getDisplayName` and `extractFirstAndLastNameFromAvailableDetails` and `getCountryISO` to PersonalDetailsUtils? They are now in actions/PersonalDetails. Should we leave in the actions only the functions that write to API?

-> TODO

> * same with  `getDisplayNameForParticipant` and other methods that deal only with allPersonalDetails object - should we move them to PersonalDetailsUtils?

-> we can make the decision after `getDisplayNameForParticipant` vs `getDisplayNameOrDefault` investigation

**Comment by koko57** [comment]
TLDR; ⬆️ 
-> I'm investigating now if we can use only one method `getDisplayNameForParticipant` or `getDisplayNameOrDefault`
-> I'll check if we can/should move `createDisplayName`, `extractFirstAndLastNameFromAvailableDetails` and `getCountryISO` to PersonalDetailsUtils

If so, I'll open a new PR, if not - we can close this issue out

**Comment by puneetlath** [comment]
Sounds great. Thanks @koko57!

**Comment by koko57** [comment]
not overdue 🙂 I'll post my findings today

**Comment by koko57** [comment]
@puneetlath regarding 

1. moving `createDisplayName`, `extractFirstAndLastNameFromAvailableDetails` and `getCountryISO` - we can safely move `createDisplayName` and`extractFirstAndLastNameFromAvailableDetails` to PersonalDetailsUtils. `getCountryISO` seems not to be used anywhere anymore, so can we remove it?
2. `getDisplayNameForParticipant` vs `getDisplayNameOrDefault` - both have 25 usages in the code, `getDisplayNameForParticipant` uses `getDisplayNameOrDefault` under the hood to get the longName https://github.com/Expensify/App/blob/2fa38fccbec0fc83a3b055fb72575a2ffc4b4ce6/src/libs/ReportUtils.ts#L1524
main differences between these two:
- we pass personalDetails (`getDisplayNameOrDefault`) instead of accountID (`getDisplayNameForParticipant`)
- in `getDisplayNameOrDefault` we display only the longName (fullName) in `getDisplayNameForParticipant` we either display the longName or the shortName
What can we do here? I think we can merge these two methods (leaving getDisplayNameOrDefault) but instead of passing personalDetails pass accountID - for most of the current version of getDisplayNameOrDefault usages we still need to get the personal details based on accountID, this logic would be moved to the refactored method. We can also pass the personalDetails additionally and check whether we passed accountID or personalDetails object.
Additionally we would need to add the param for displaying short/long name and add the logic to do so.

Let me know what do you think about it 🙂 
Moving the methods and removing the unused one is not a big deal, but merging these two methods is a bigger refactor.

**Comment by puneetlath** [comment]
I think all of that sounds good. What do you think @situchan?

**Comment by situchan** [comment]
Sounds good to me too. Thanks

**Comment by puneetlath** [comment]
@situchan how's this going?

**Comment by situchan** [comment]
review is in progress. will complete today

**Comment by puneetlath** [comment]
What's left here?

**Comment by koko57** [comment]
@puneetlath I suggested to merge getDisplayNameForParticipant with getDisplayNameOrDefault. It's a big change, I'm still looking if there won't be any edge cases and we can use only one method everywhere.

**Comment by puneetlath** [comment]
Ok, sounds good. Thanks for keeping me in the loop.

**Comment by koko57** [comment]
@puneetlath After analyzing the usages of both methods I'm not convinced that we should go with merging them - they are slightly different and I would need to introduce an additional layer of complexity to cover all the cases. For now, I think that both methods serve their purpose in the places they are used and are easy to use. 
If you agree, we can close this issue.

**Comment by puneetlath** [comment]
Ok sounds good to me. @situchan do you agree? 

Thanks for all your effort on this @koko57!

**Comment by situchan** [comment]
Agree. Thanks

**Comment by situchan** [comment]
- #31495
- #33930
- #34657

**Comment by puneetlath** [comment]
Great, thanks. Offer sent: https://www.upwork.com/nx/wm/offer/100915132

Please ping me here when you've accepted.

**Comment by situchan** [comment]
Accepted thanks

**Comment by puneetlath** [comment]
Paid. Thanks y'all!
